schemachange/mixed-versions: handle possible syntax error for BIT(0) column usage

spilchen opened the pull request:
A recent fix enabled support for BIT(0) columns (see issue #132944), which has since been backported. However, since this backport hasn’t been released yet, running against an older down-level client may still trigger a syntax error. This change adds a temporary check to handle such errors until the backports for #132944 are publicly released.

This adjustment will be backported to older branches, with slight variations per release:
- release-24.3: Reuse the same fix as in master.
- release-24.2: Use isClusterVersionLessThan with clusterversion.V24.2.
- release-24.1: Use isClusterVersionLessThan with clusterversion.V24.1.
- release-23.2: Use isClusterVersionLessThan with clusterversion.V23.2.

Epic: None
Release note: None
Closes #133339
Closes #133790
Closes #131162

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/134407)
<!-- Reviewable:end -->

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/134407)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong, @spilchen, and @vidit-bhat)

___
*[`pkg/workload/schemachange/operation_generator.go` line 1281 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/134407#-OB1nKM89A_d3SXF8sEf:-OB1nKM9Ej1TnyATsZSo:b-fttct2) ([raw file](https://github.com/cockroachdb/cockroach/blob/ffdfb4062a5a5463e7a63dd2769dda151dac0ee9/pkg/workload/schemachange/operation_generator.go#L1281)):*
> ```Go
> 		return false
> 	}()
> 	hasBit0Type := func() bool {
> ```

i think we'll want to block this in other places also, like in ADD COLUMN and SET DEFAULT

perhaps we could add one helper function where we check if the type is allowed so we can use it in a few places.


<!-- Sent from Reviewable.io -->

rafiss commented on pkg/workload/schemachange/operation_generator.go:
```diff
@@ -4808,6 +4832,22 @@ func isClusterVersionLessThan(
+// isMixedVersionTest works similarly to isClusterVersionLessThan, but without
+// specifying a version. It returns true if the cluster version is not the
+// latest, indicating a mixed-version test.
+func isMixedVersionTest(ctx context.Context, tx pgx.Tx) (bool, error) {
```
super nit: it could be called `inMixedVersionState`, since technically the whole test could be considered a "mixed version test"

rafiss reviewed (approved):
nice! thanks for the fix

spilchen committed:
schemachange/mixed-versions: handle possible syntax error for BIT(0) column usage

A recent fix enabled support for BIT(0) columns (see issue #132944),
which has since been backported. However, since this backport hasn’t
been released yet, running against an older down-level client may still
trigger a syntax error. This change adds a temporary check to handle
such errors until the backports for #132944 are publicly released.

This adjustment will be backported to older branches, with slight
variations per release:
- release-24.3: Reuse the same fix as in master.
- release-24.2: Use isClusterVersionLessThan with clusterversion.V24.2.
- release-24.1: Use isClusterVersionLessThan with clusterversion.V24.1.
- release-23.2: Use isClusterVersionLessThan with clusterversion.V23.2.

Epic: None
Release note: None
Closes: #133339, #133790, #131162

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/134407)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong, @rafiss, and @vidit-bhat)

___
*[`pkg/workload/schemachange/operation_generator.go` line 1281 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/134407#-OB1nKM89A_d3SXF8sEf:-OB2GSWm9HH9tFvIFkMD:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/ffdfb4062a5a5463e7a63dd2769dda151dac0ee9/pkg/workload/schemachange/operation_generator.go#L1281)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i think we'll want to block this in other places also, like in ADD COLUMN and SET DEFAULT

perhaps we could add one helper function where we check if the type is allowed so we can use it in a few places.
</blockquote></details>

Done
___
*[`pkg/workload/schemachange/operation_generator.go` line 4838 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/134407#gh-1831688139:-OB2GXM0565814l2t7wJ:bdcv6l8) ([raw file](https://github.com/cockroachdb/cockroach/blob/5750fb9519100a8f651d43629ed0bab0adf82c35/pkg/workload/schemachange/operation_generator.go#L4838)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

super nit: it could be called `inMixedVersionState`, since technically the whole test could be considered a "mixed version test"
</blockquote></details>

Good point.


<!-- Sent from Reviewable.io -->

spilchen commented:
TFTR!

bors r+
